enzyme/Enzyme/MLIR/Analysis/Lattice.h

```diff
@@ -0,0 +1,364 @@
+
+  SetLattice(ValueT single) : state(State::Defined) { elements.insert(single); }
+
+  // TODO(zinenko): deprecate this and use a visitor instead.
+  DenseSet<ValueT> &getElements() {
+    assert(state == State::Defined);
+    return elements;
+  }
+
+  const DenseSet<ValueT> &getElements() const {
+    return const_cast<SetLattice<ValueT> *>(this)->getElements();
+  }
+
+  bool isUnknown() const { return state == State::Unknown; }
+  bool isUndefined() const { return state == State::Undefined; }
+
+  ChangeResult join(const SetLattice<ValueT> &other) {
+    if (isUnknown())
+      return ChangeResult::NoChange;
+    if (isUndefined() && other.isUndefined())
+      return ChangeResult::NoChange;
+    if (other.isUnknown()) {
+      state = State::Unknown;
+      return ChangeResult::Change;
+    }
+
+    ChangeResult result = updateStateToDefined();
+    return insert(other.elements) | result;
+  }
+
+  ChangeResult insert(const DenseSet<ValueT> &newElements) {
+    if (isUnknown())
+      return ChangeResult::NoChange;
+
+    size_t oldSize = elements.size();
+    elements.insert(newElements.begin(), newElements.end());
+    ChangeResult result = elements.size() == oldSize ? ChangeResult::NoChange
+                                                     : ChangeResult::Change;
+    return updateStateToDefined() | result;
+  }
+
+  ChangeResult markUnknown() {
+    if (isUnknown())
+      return ChangeResult::NoChange;
+
+    state = State::Unknown;
+    elements.clear();
+    return ChangeResult::Change;
+  }
+
+  /// Returns true if this set is in the canonical form, i.e. either the state
+  /// is `State::Defined` or the explicit list of classes is empty, but not
+  /// both.
+  bool isCanonical() const {
+    return state == State::Defined || elements.empty();
+  }
+
+  /// Returns an instance of SetLattice known not to have any elements.
+  /// This is different from "undefined" and "unknown". The instance is *not* a
+  /// classical singleton.
+  static const SetLattice<ValueT> &getEmpty() {
+    static const SetLattice<ValueT> empty(State::Defined);
+    return empty;
+  }
+
+  /// Returns an instance of SetLattice in "undefined" state, i.e. without a set
+  /// of elements. This is different from empty set, which indicates that the
+  /// set is known not to contain any elements. The instance is *not* a
+  /// classical singleton, there are other ways of obtaining it.
+  static const SetLattice<ValueT> &getUndefined() { return undefinedSet; }
+
+  /// Returns an instance of SetLattice for the "unknown" class. The instance
+  /// is *not* a classical singleton, there are other ways of obtaining an
+  /// "unknown" alias set.
+  static const SetLattice<ValueT> &getUnknown() { return unknownSet; }
+
+  bool operator==(const SetLattice<ValueT> &other) const {
+    assert(isCanonical() && other.isCanonical());
+    return state == other.state && llvm::equal(elements, other.elements);
+  }
+
+  void print(llvm::raw_ostream &os) const {
```

Comment:
```suggestion
  LLVM_DUMP_METHOD void print(llvm::raw_ostream &os) const {
```

Add LLVM_DUMP_METHOD so that this method does not get stripped for debug builds.